From 5dabfbfe0398d12f97b14f7144d671868ecf59e5 Mon Sep 17 00:00:00 2001
From: Arnaud Pouliquen <arnaud.pouliquen@foss.st.com>
Date: Mon, 20 Jun 2022 13:47:32 +0200
Subject: [PATCH] Revert "remoteproc: stm32: use workqueue to treat mailbox
 callback"

This reverts commit 714cf5e3846047052fae89c7b927b886cc69d60c.

For the rpmsg_i2c and rpmsg_intc drivers support we need to manage the
RPMsg in interrupt context.
This revert suppresses the creation of a workqueue, which
allowed to call the RPMsg framework in normal context.

Change-Id: Ia60c78ee416904b21504123e225d152daa3aea02
Signed-off-by: Arnaud Pouliquen <arnaud.pouliquen@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/286891
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/347948
Domain-Review: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
Tested-by: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
Reviewed-by: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 drivers/remoteproc/stm32_rproc.c | 45 ++++----------------------------
 1 file changed, 5 insertions(+), 40 deletions(-)

--- a/drivers/remoteproc/stm32_rproc.c
+++ b/drivers/remoteproc/stm32_rproc.c
@@ -21,7 +21,6 @@
 #include <linux/reset.h>
 #include <linux/slab.h>
 #include <linux/tee_remoteproc.h>
-#include <linux/workqueue.h>
 
 #include "remoteproc_internal.h"
 
@@ -35,9 +34,7 @@
 #define STM32_SMC_REG_WRITE	0x1
 
 #define STM32_MBX_VQ0		"vq0"
-#define STM32_MBX_VQ0_ID	0
 #define STM32_MBX_VQ1		"vq1"
-#define STM32_MBX_VQ1_ID	1
 #define STM32_MBX_SHUTDOWN	"shutdown"
 #define STM32_MBX_DETACH	"detach"
 
@@ -77,7 +74,6 @@ struct stm32_mbox {
 	const unsigned char name[10];
 	struct mbox_chan *chan;
 	struct mbox_client client;
-	struct work_struct vq_work;
 	int vq_id;
 };
 
@@ -92,7 +88,6 @@ struct stm32_rproc {
 	u32 nb_rmems;
 	struct stm32_rproc_mem *rmems;
 	struct stm32_mbox mb[MBOX_NB_MBX];
-	struct workqueue_struct *workqueue;
 	bool hold_boot_smc;
 	bool fw_loaded;
 	struct tee_rproc *trproc;
@@ -433,30 +428,13 @@ static irqreturn_t stm32_rproc_wdg(int i
 	return IRQ_HANDLED;
 }
 
-static void stm32_rproc_mb_vq_work(struct work_struct *work)
-{
-	struct stm32_mbox *mb = container_of(work, struct stm32_mbox, vq_work);
-	struct rproc *rproc = dev_get_drvdata(mb->client.dev);
-
-	mutex_lock(&rproc->lock);
-
-	if (rproc->state != RPROC_RUNNING && rproc->state != RPROC_ATTACHED)
-		goto unlock_mutex;
-
-	if (rproc_vq_interrupt(rproc, mb->vq_id) == IRQ_NONE)
-		dev_dbg(&rproc->dev, "no message found in vq%d\n", mb->vq_id);
-
-unlock_mutex:
-	mutex_unlock(&rproc->lock);
-}
-
 static void stm32_rproc_mb_callback(struct mbox_client *cl, void *data)
 {
 	struct rproc *rproc = dev_get_drvdata(cl->dev);
 	struct stm32_mbox *mb = container_of(cl, struct stm32_mbox, client);
-	struct stm32_rproc *ddata = rproc->priv;
 
-	queue_work(ddata->workqueue, &mb->vq_work);
+	if (rproc_vq_interrupt(rproc, mb->vq_id) == IRQ_NONE)
+		dev_dbg(&rproc->dev, "no message found in vq%d\n", mb->vq_id);
 }
 
 static void stm32_rproc_free_mbox(struct rproc *rproc)
@@ -474,7 +452,7 @@ static void stm32_rproc_free_mbox(struct
 static const struct stm32_mbox stm32_rproc_mbox[MBOX_NB_MBX] = {
 	{
 		.name = STM32_MBX_VQ0,
-		.vq_id = STM32_MBX_VQ0_ID,
+		.vq_id = 0,
 		.client = {
 			.rx_callback = stm32_rproc_mb_callback,
 			.tx_block = false,
@@ -482,7 +460,7 @@ static const struct stm32_mbox stm32_rpr
 	},
 	{
 		.name = STM32_MBX_VQ1,
-		.vq_id = STM32_MBX_VQ1_ID,
+		.vq_id = 1,
 		.client = {
 			.rx_callback = stm32_rproc_mb_callback,
 			.tx_block = false,
@@ -538,10 +516,6 @@ static int stm32_rproc_request_mbox(stru
 			dev_warn(dev, "cannot get %s mbox\n", name);
 			ddata->mb[i].chan = NULL;
 		}
-		if (ddata->mb[i].vq_id >= 0) {
-			INIT_WORK(&ddata->mb[i].vq_work,
-				  stm32_rproc_mb_vq_work);
-		}
 	}
 
 	return 0;
@@ -1022,18 +996,12 @@ static int stm32_rproc_probe(struct plat
 		rproc->state = RPROC_DETACHED;
 
 	rproc->has_iommu = false;
-	ddata->workqueue = create_workqueue(dev_name(dev));
-	if (!ddata->workqueue) {
-		dev_err(dev, "cannot create workqueue\n");
-		ret = -ENOMEM;
-		goto free_resources;
-	}
 
 	platform_set_drvdata(pdev, rproc);
 
 	ret = stm32_rproc_request_mbox(rproc);
 	if (ret)
-		goto free_wkq;
+		goto free_resources;
 
 	ret = rproc_add(rproc);
 	if (ret)
@@ -1043,8 +1011,6 @@ static int stm32_rproc_probe(struct plat
 
 free_mb:
 	stm32_rproc_free_mbox(rproc);
-free_wkq:
-	destroy_workqueue(ddata->workqueue);
 free_resources:
 	rproc_resource_cleanup(rproc);
 free_rproc:
@@ -1071,7 +1037,6 @@ static void stm32_rproc_remove(struct pl
 
 	rproc_del(rproc);
 	stm32_rproc_free_mbox(rproc);
-	destroy_workqueue(ddata->workqueue);
 
 	if (device_may_wakeup(dev)) {
 		dev_pm_clear_wake_irq(dev);
